### PR TITLE
[darwin] Update BLEScannerDelegate interface to have both a add and a…

### PR DESCRIPTION
**Pull request by vivien-apple**
… remove method

#### Problem

There is no mechanism to know when a peripheral has been removed.

### CODE REVIEW

File: src/platform/Darwin/BleConnectionDelegateImpl.mm [hunk 1]
```diff
@@ -630,6 +622,8 @@ - (void)updateWithDiscriminator:(const chip::SetupDiscriminator &)deviceDiscrimi
         }
     }
 
+    [self removePeripheralsFromCache];
```
Comment (bzbarsky-apple) [inline src/platform/Darwin/BleConnectionDelegateImpl.mm:625]: This nulls out `_cachedPeripherals`.  Is that expected?

File: src/platform/Darwin/BleConnectionDelegateImpl.mm [hunk 2]
```diff
@@ -649,6 +643,64 @@ - (void)updateWithPeripheral:(CBPeripheral *)peripheral
     [self stopScanning];
 }
 
+- (void)addPeripheralToCache:(CBPeripheral *)peripheral data:(NSData *)data
+{
+    dispatch_source_t timeoutTimer;
```
Comment (bzbarsky-apple) [inline src/platform/Darwin/BleConnectionDelegateImpl.mm:648]: Should this bail out early if `_cachedPeripherals` is nil?  Or is it never expected to be nil here?

File: src/platform/Darwin/BleConnectionDelegateImpl.mm (hunk 1 above)
Comment (vivien-apple) [reply src/platform/Darwin/BleConnectionDelegateImpl.mm:625]: No that was not! Thanks !

File: src/platform/Darwin/BleConnectionDelegateImpl.mm (hunk 2 above)
Comment (vivien-apple) [reply src/platform/Darwin/BleConnectionDelegateImpl.mm:649]: It is not supposed to be nil. (which was a wrong assertion with `removePeripheralsFromCache` as you mentioned).

**Review by bzbarsky-apple — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Pull request merged**